doc/elements.md

```diff
@@ -78,6 +78,33 @@ Attribute | Type | Default | Description
 `digits` | integer | 2 | number of digits that must be correct for `comparison="sigfig"` or `comparison="decdig"`.
 `eps_digits` | integer | 3 | Additional digits (beyond `digits`) used to compute a grace tolerance.
 
+## `pl_file_download` element
+
+```html
+<!-- allow students to download an existing file -->
+<pl_file_download file_name="data.txt" directory="clientFilesCourse" />
+
+<!-- allow students to download a file that is generated by code -->
+<pl_file_download file_name="data.txt" type="dynamic" />
+```
+
+Attribute | Type | Default | Description
+--- | --- | --- | ---
+`file_name` | string | — | Name of file to download.
+`label` | text | file_name | Alternate text for file download link (e.g., `label="click here to download"`).
+`type` | text | "static" | Type of file, either "static" (an existing file) or "dynamic" (a file generated by element or server code).
+`directory` | text | "clientFilesQuestion" | The directory that contains the file, either "clientFilesQuestion" or "clientFilesCourse" (see [client and server files](clientServerFiles.md)). A directory cannot be specified if `type="dynamic"`.
+
+If `type="dynamic"`, then the contents of the file must be returned by a function `file()` that is located either in element code or in `server.py`. The contents must be a string (with utf-8 encoding), a bytes-like object, or a file-like object. The filename will be available to this function as `data['filename']`. For example, this code might appear in `server.py` to generate a file called `data.txt`:
+
+```python
+def file(data):
+    if data['filename']=='data.txt':
+        return 'This data is generated by code.'
+    else:
```

Comment:
Is this `else:` necessary? If we caught `None` as a return value in the trampoline and treated it as `''` then we could let `server.py` just not return anything if `filename` doesn't match.